### PR TITLE
[PlaygroundLogger] Implemented a check to avoid trying to log nil-con…

### PR DESCRIPTION
**Pull request by cwakamo**
…taining non-Optional objects.

It is undefined behavior for a non-Optional object to be nil in Swift, and at present, PlaygroundLogger triggers crashes in the runtime when trying to log such objects.
The most common way for such an object to exist is for an Objective-C API to be annotated as returning nonnull but in fact returns nil.
In a regular Swift program, this can be worked around by explicitly assigning the value to an Optional.
PlaygroundLogger cannot do that as it only receives an Any, and assigning to Any? does not recover the nil.

As a result, inspect the contents of the Any's existential container (i.e. the thing in memory that is the Any itself, not the value inside of the Any) to determine if it contains a non-Optional object whose value is nil.

This commit includes additional tests to ensure that we can log this broken value without crashing and that we can log a nil Optional object correctly.

This addresses <rdar://problem/56370098>.

### CODE REVIEW

File: PlaygroundLogger/PlaygroundLogger/LogEntry+Reflection.swift [hunk 1]
```diff
@@ -31,6 +31,32 @@ extension LogEntry {
             return
         }
 
+        // In Swift, it is undefined behavior to have a value of non-Optional class type which is set to nil.
+        // Unfortunately, these can be encountered in the wild; the most common example is an incorrect Objective-C nullability annotation.
+        // To prevent PlaygroundLogger from trigger undefined behavior (which, as of this writing, causes a crash in the Swift runtime), perform a safety check before proceeding with logging.
```
Comment (jonathanpenn) [inline PlaygroundLogger/PlaygroundLogger/LogEntry+Reflection.swift]: "from trigger undefined" seems like it should be "from triggering undefined".

File: PlaygroundLogger/PlaygroundLoggerTests/PGLNilObject.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface PGLNilObject : NSObject
```
Comment (jonathanpenn) [inline PlaygroundLogger/PlaygroundLoggerTests/PGLNilObject.h]: Would it be useful to annotate this type to explain what its for? I know you did that in the `testNonOptionalNilObject()` method above, but it seems like putting it here might make sense.

**Comment by cwakamo** [comment]
@jonathanpenn I made the changes you requested, would you mind taking another look at this PR?

**Comment by jonathanpenn** [comment]
Looks great!

**Pull request merged**